### PR TITLE
feat(deployer): add more deployment metadata

### PR DESCRIPTION
**Pull request by hseeberger**
Ref #941.

### CODE REVIEW

File: cargo-shuttle/src/lib.rs [hunk 1]
```diff
@@ -901,14 +901,51 @@ impl Shuttle {
     }
 
     async fn deploy(&self, client: &Client, args: DeployArgs) -> Result<CommandOutcome> {
-        if !args.allow_dirty {
-            self.is_dirty()?;
+        let working_directory = self.ctx.working_directory();
+
+        let mut deploy_options: HashMap<String, String> = HashMap::new();
```
Comment (hseeberger) [inline cargo-shuttle/src/lib.rs:906]: I think we can omit this type annotation if we use `to_string` in line 922 instead of `into`.

Comment (hseeberger) [reply cargo-shuttle/src/lib.rs:906]: > I prefer the explicit annotation in this case.
Really? I try to always avoid them. Taste is different.

Comment (hseeberger) [reply cargo-shuttle/src/lib.rs:906]: > However, I think we should consider using a serde struct instead of a HashMap ...

Type safety is always good! We could also use a "simple" (no extractor) struct like this:

``` rust
struct DeployOptions {
    git_branch: Option<String>,
    git_dirty: bool,
    ...
}
```

**Comment by hseeberger** [comment]
I really like the `DeploymentRequest` "struct approach".

**Review by iulianbarbu — COMMENTED** [review]
This is almost there @hseeberger ! Thanks a lot for starting this, it is an important project for our upcoming console.

File: cargo-shuttle/src/lib.rs [hunk 2]
```diff
@@ -901,15 +901,43 @@ impl Shuttle {
...
+
+            if let Ok(head) = repo.head() {
+                // This is typically the name of the current branch
+                // It is "HEAD" when head detached, for example when a tag is checked out
+                deploy_req.git_branch = head.shorthand().map(String::from);
+                if let Ok(commit) = head.peel_to_commit() {
+                    let commit_id = commit.id();
+                    deploy_req.git_commit_id = Some(commit_id.to_string());
+                    // message is valid utf-8
+                    deploy_req.git_commit_msg = commit.message().map(String::from);
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply cargo-shuttle/src/lib.rs:932]: Yeah, we are going to trim it in some way.

File: deployer/src/persistence/deployment.rs [hunk 3]
```diff
@@ -17,6 +18,10 @@ pub struct Deployment {
     pub last_update: DateTime<Utc>,
     pub address: Option<SocketAddr>,
     pub is_next: bool,
+    pub git_commit_id: Option<String>,
+    pub git_commit_msg: Option<String>,
+    pub git_branch: Option<String>,
+    pub git_dirty: bool,
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/persistence/deployment.rs:24]: The default value None will map to false in later stages anyways

File: deployer/src/handlers/mod.rs [hunk 4]
```diff
@@ -306,37 +305,34 @@ pub async fn create_service(
...
     Extension(claim): Extension<Claim>,
     Path((project_name, service_name)): Path<(String, String)>,
-    Query(params): Query<HashMap<String, String>>,
-    mut stream: BodyStream,
+    Json(body): Json<DeploymentRequest>,
 ) -> Result<Json<shuttle_common::models::deployment::Response>> {
     let service = persistence.get_or_create_service(&service_name).await?;
     let id = Uuid::new_v4();
 
+    let data = todo!("Base64 decode body.archive");
```
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/handlers/mod.rs:313]: Do you think base64 in a JSON body is a suitable strategy?

[the remark this answers is not in the document]
Comment (iulianbarbu) [reply deployer/src/handlers/mod.rs:313]: Good question. I would use `bincode` especially because of being compatible with serde. I am not sure how easily would be to have the ser/deser done with `base64`. What do you think?

File: deployer/src/persistence/deployment.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply deployer/src/persistence/deployment.rs:24]: Hmm, what do you mean by later stages? I am wondering if a NULL (for the old deployments) will map to `false` or will result into panic.

File: cargo-shuttle/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (hseeberger) [reply cargo-shuttle/src/lib.rs:932]: I suggest using `commit.summary()`.

File: deployer/src/persistence/deployment.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/persistence/deployment.rs:24]: Hmm, yeah I guess that's what I meant. NULL will probably be interpreted as "not dirty", but the distinction between NULL and false is reasonable. It should be an Option.

File: deployer/src/handlers/mod.rs [hunk 5]
```diff
@@ -306,37 +305,34 @@ pub async fn create_service(
...
+
     let deployment = Deployment {
         id,
         service_id: service.id,
         state: State::Queued,
         last_update: Utc::now(),
         address: None,
         is_next: false,
+        git_commit_id: body.git_commit_id,
+        git_commit_msg: body.git_commit_msg,
```
[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs:323]: Hmm ... might make sense, yet then we maybe also should trim the commit_id, just in case the caller (maybe maliciously) passes a very long string value.

To be super clean, we should use new types like `CommitId` and `CommitMsg` and create associated factories which take a `String` and return a `Result<Self, Error>` in case some validations fail.

I like that approach, but I am not sure if this is in sync with the overall implementation patterns / coding style of Shuttle. Of course we could try to be a shining example.

File: deployer/src/handlers/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs:313]: Will be looking into bincode encoding.

File: deployer/src/handlers/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs:323]: @iulianbarbu, should we just use `commit.summary()` and leave more sphisticated type foo for later?

File: cargo-shuttle/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonaro00) [reply cargo-shuttle/src/lib.rs:934]: It seems like --abbrev-commit is equivalent to `commit.summary()`.
It takes the first paragraph and collapses onto one line.

File: deployer/src/handlers/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/handlers/mod.rs:323]: If there was a max limit on commit msg (summary) length, and a user happens to have a very long one, I think they shouldn't be denied to deploy because of that error.
Just trimming it gives shuttle the info it needs.

[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs:323]: But `summary` is not fallible. It just gets "the short “summary” of the git commit message."

File: cargo-shuttle/src/lib.rs [hunk 6]
```diff
@@ -901,15 +901,45 @@ impl Shuttle {
...
+
+            if let Ok(head) = repo.head() {
+                // This is typically the name of the current branch
+                // It is "HEAD" when head detached, for example when a tag is checked out
+                deploy_req.git_branch = head.shorthand().map(String::from);
+                if let Ok(commit) = head.peel_to_commit() {
+                    let commit_id = commit.id();
+                    deploy_req.git_commit_id = Some(commit_id.to_string());
+                    // message is valid utf-8
+                    deploy_req.git_commit_msg = commit.summary().map(|s| s.chars().take(80).collect());
```
Comment (jonaro00) [inline cargo-shuttle/src/lib.rs:934]: I implemented string trimming on both ends like this.
Ideally it would trim all incoming strings that are going to be put into a DB.
What do you think? (will use shared constant if approved)

File: deployer/src/handlers/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonaro00) [reply deployer/src/handlers/mod.rs]: It can still be very *not* short :D 
(`commit.summary()` gives the same string)
image

File: cargo-shuttle/src/lib.rs (hunk 6 above)
Comment (hseeberger) [reply cargo-shuttle/src/lib.rs:934]: Looking good.

File: deployer/src/handlers/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs]: Oh!

File: deployer/src/handlers/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply deployer/src/handlers/mod.rs:318]: @hseeberger , @oddgrd mentioned to me that we already use `rmp-serde` for `shuttle-next`. Have you managed to use `bincode` already? I think it is at least the more popular encoding/decoding solution in the Rust ecosystem. There are figures that support it: [1] and [2], but there is a limitation as well [3]. I still think it's worth using it for Shuttle. The `self-describing` aspect described at [3] shouldn't be a must for us.

[1] https://github.com/djkoloski/rust_serialization_benchmark#raw-data
[2] https://users.rust-lang.org/t/choice-of-format-for-serialisation/81339 - seems like `rmp-serde` brings some significant overhead when serializing `Vec<u8>` for a specific case of a user. I am concerned about this because we tend to fit the same usage description when sending our code source archive in the body. A bit concerning for a big scale down the line, sending that much self-describing data over the network, which is not really needed when deserializing a static type.
[3] `bincode` doesn't support dynamically typed deserialization. I think this shouldn't be a problem for us if we provide the static type information when deserializing, rather than using an approach similar to `serde_json` [4], which can infer the types by seeking through the input for key parts (`"` for the beginning of a string, `{` for the beginning of an object etc.)
[4] https://serde.rs/impl-deserialize.html

[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs:318]: So you suggest that we use rmp-serde instead of bincode? Fine for me, I‘ll take a look.

[the remark this answers is not in the document]
Comment (oddgrd) [reply deployer/src/handlers/mod.rs:318]: I believe @iulianbarbu  is suggesting using bincode, he and I discussed this earlier and we landed on that. I think it's a good choice, especially if you have already started implementing it. However, after thinking about this some more I am not as certain, the consideration here is if we want to optimize for serialization/deserialization speed (bincode should be better) or the size of the serialized data (msgpack should be better). Smaller sizes might be what we want to optimize for (users will have smaller uploads). To get a proper understanding of the performance of the two for our usecase, it would be good to test both against a shuttle project and see how they fare (I wouldn't call that a requirement of this PR though).

Either way there are a lot of serialization formats in rust, and getting bogged down in optimizing this right now may be premature. It shouldn't be too much trouble to switch it out later. 

TLDR: feel free to proceed with bincode, and we can measure the performance of the alternatives later and optimize it if needed.

[the remark this answers is not in the document]
Comment (iulianbarbu) [reply deployer/src/handlers/mod.rs:318]: Sorry for the confusion. I was talking about using `bincode` instead of `rmp-serde` because of the simplicity and ser/deser speed, but have some reasons to think it doesn't matter which one we use. 

There is a question about the serialization size and the overhead of sending the serialized buffer over the network. `rmp-serde` seems to do better for small inputs size in terms of space (per [1]). Shuttle projects deployment metadata is fixed but the source code archive varies in size, especially when including assets folders. The big inputs might be handled better by `bincode` (speed and space-wise, per [1] again).  I think though that the majority of the projects are small sized and we're not that concerned right now about the overheads of sending deployment info over the network + ser/deser, but rather on the project build times, which take up the majority of the deployment time.

The end conclusion (at least for me) is that at this moment we don't have enough data to support using `bincode` over `rmp-serde`, or vice versa. We can stick to using `rmp-serde` to not include another dependency like `bincode` if you haven't made significant progress with `bincode`. I think doing a comparison for ser/deser speeds of Shuttle-like inputs (deployment metadata + source code archive of a Shuttle example) might help guide our decision a bit, but still feels like overengineering to me at this stage. Also, going with one encoding and ser/deser (serde based) library is not a one-way door, so I don't want to weigh it that heavily. Feel free to go with `rmp-serde` now if that's fine for you.

File: deployer/src/handlers/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply deployer/src/handlers/mod.rs]: @hseeberger @jonaro00 , seems like you guys discovered `commit.summary()` is not enough. I think trimming it to 80 chars in addition should do it though.

File: cargo-shuttle/src/lib.rs (hunk 6 above)
Comment (iulianbarbu) [reply cargo-shuttle/src/lib.rs]: sgtm

File: deployer/src/handlers/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs:318]: No worries, guys, either way should be fine. Given the state of the discussion, I think that "not include another dependency" is probably the most compelling reason to use rmp-serde. Anyway, I have a feeling that implementing this feature will take less effort than this discussion so far ;-)

The reason it has not happened yet is that I have had limited time due to my mother's health issues and vacation. But I will start working on this RP (again) soon.

[the remark this answers is not in the document]
Comment (oddgrd) [reply deployer/src/handlers/mod.rs:318]: No problem, take your time!

[the remark this answers is not in the document]
Comment (hseeberger) [reply deployer/src/handlers/mod.rs]: Guys, I have implemented the encoding and decoding of the newly introduced `DeploymentRequest` via `rmp_serde`. Please take a look.

Notice that for decoding I introduced the custom Axum extractor `Rmp`. Curretnly located in `delpoyer::handlers`, I guess we could lift it to the `common` crate or somewhere else where it can be reused (after adding docs).

File: cargo-shuttle/src/lib.rs [hunk 7]
```diff
@@ -908,14 +910,48 @@ impl Shuttle {
...
+                .context("getting working directory of repository")?;
+            let repo_path = dunce::canonicalize(repo_path)?;
+            trace!(?repo_path, "found git repository");
+
+            let is_dirty = self.is_dirty(&repo).err();
+            if !args.allow_dirty {
+                if let Some(err) = is_dirty {
+                    bail!(err);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (hseeberger) [reply cargo-shuttle/src/lib.rs:932]: Thanks for the suggestion. I think we can write it even more concise and clearer:

```
            if !args.allow_dirty {
                self.is_dirty(&repo).context("dirty not allowed")?;
            }
```

File: cargo-shuttle/src/lib.rs [hunk 8]
```diff
@@ -908,14 +910,48 @@ impl Shuttle {
...
+            deployment_req.git_dirty = Some(is_dirty.is_some());
```
[the remark this answers is not in the document]
Comment (hseeberger) [reply cargo-shuttle/src/lib.rs:933]: Of course! With my above suggestion: 

```
deployment_req.git_dirty = Some(self.is_dirty(&repo).is_err());
```

File: cargo-shuttle/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply cargo-shuttle/src/lib.rs]: This is great! Thanks.

**Review by iulianbarbu — COMMENTED** [review]
Thanks! The data is handled well for both old and new projects. The only thing I noticed is that when doing `cargo-shuttle deployments list` we do not get the git metadata in the table. I thought it would be useful to be shown there, otherwise this info will be hidden to the CLI users.

**Comment by jonaro00** [comment]
@iulianbarbu Adding a way to fetch this metadata was not in spec for this PR. What you describe should of course be added, but we can keep that in a separate issue.

**Comment by iulianbarbu** [comment]
> @iulianbarbu Adding a way to fetch this metadata was not in spec for this PR. What you describe should of course be added, but we can keep that in a separate issue.

Indeed. Would be great to open an issue to track it then. 👍

**Review by iulianbarbu — APPROVED** [review]
LGTM and thanks again @hseeberger and @jonaro00 !

**Review by oddgrd — APPROVED** [review]
LGTM too, and great work making this a non-breaking change!
